### ISSUE TITLE
Tool Should be OS Sensitive 

### ISSUE THREAD

**Issue by amit-traiana** [rules-proto-grpc/rules_proto_grpc#44]
Hi again :-)

We tried running our Bazel project on Windows, and I'm getting the following error when trying to compile `scala_grpc_*` targets: 

```
--grpc_scala_plugin_out: protoc-gen-grpc_scala_plugin: %1 is not a valid Win32 application.
```

Had a look on the code, and it looks like `@com_github_scalapb_scalapb` reference a zip file (`scalapbc-0.9.4.zip`). The zip file has a `bin` folder that has two `proto-gen-scala` files, one binary for Windows, and one for Linux. `scalapb` [BUILD](https://github.com/rules-proto-grpc/rules_proto_grpc/blob/master/third_party/BUILD.bazel.com_github_scalapb_scalapb) file only exports the Linux file.

Also, the tool defined in the Scala BUILD file should be OS Sensitive. I'm assuming the easiest way to fix it, is to add another parameter to `proto_plugin` called `tool_windows` and set a default value to it. Then - `proto_plugin_impl` can check if the parameters is not set to the default value, and if indeed the OS is Windows, and if so - set the ProtoPluginInfo Provider tool to be the windows tool (or else just the `tool`). This will have the least impact on other BUILD files, because everything can just keep working the same way it was - and `proto_plugin_impl` will just pick the the `tool` parameter normally. 

If you don't want another parameter you can change `tool` to be an array of sort that has two tools, but this means a broader change across the project. 

Let me know if this makes sense for you - or perhaps you got a better idea. If needed, I might be able to help with a Pull Request.

**Comment by aaliddell** [comment]
This might be solvable with a select on the platform

**Comment by amit-traiana** [comment]
I think you already importing `bazel-tools`? If so, I think they have `@bazel_tools//src/conditions:windows`. I'll add the tool the scalapb BUILD file, and try give `select` a try tomorrow.

**Comment by aaliddell** [comment]
It shouldn't even need bazel-tools: `@platforms//os:windows` should work I think

**Comment by amit-traiana** [comment]
Thanks for the heads up. The problem with using `@platform//os:windows` is that you have to create you in-house constraints. There's seems to be a simpler way that is used by `bazel_tool`:

```
    tool = select({"@bazel_tools//src/conditions:host_windows": "@com_github_scalapb_scalapb//:bin/protoc-gen-scala.bat",
                   "//conditions:default": "@com_github_scalapb_scalapb//:bin/protoc-gen-scala"}),
```

There's seems to be a `protoc` problem associated with that though. `protoc` can accept "*nix" style folder structure in Windows. for example:

```
 --descriptor_set_in=bazel-out/k8-fastbuild/bin/external/com_google_protobuf/descriptor_proto-descriptor-set.proto.bin`
```

The above will work, even though in Windows everything has to be with `\` instead. But for some reason, the same behaviour won't be applied to `--plugin` option.

```
-plugin=protoc-gen-grpc_scala_plugin=external\com_github_scalapb_scalapb\bin\protoc-gen-scala.bat 
```

Will output an error:
```
'external' is not recognized as an internal or external command,
operable program or batch file
```

It seems like it's getting confused. I manually changed the *nix slashes into windows, and it compiles successfully. So before I'll open a PR, I want to figure out how to handle this issue. Have you encountered it before? Adding stuff like `"` to the folder name didn't work.

**Comment by amit-traiana** [comment]
There was some [talk about](https://github.com/bazelbuild/bazel/issues/2045) it on the Bazel repo. It looks like it either:

1. Smuggle an 'is_windows' argument (which will be defined in `proto_plugin`) to `aspect.bzl` and create the `--plugin` folder structure based on that.
2. Use the [host_path_separator](https://docs.bazel.build/versions/master/skylark/lib/configuration.html#host_path_separator) hack mentioned there. If it return `;`, it means it's a Windows machine - and the path can be adjusted. 
3. Implemented a tool-chain per OS, which is probably the "Right" way, but this pretty much break the current design of proto_plugin - correct me if I'm wrong. 

Let me know what you think.

**Comment by aaliddell** [comment]
Whilst 3 sounds like the proper way for capturing OS specific traits, if it's just for mangling a single path then I'm inclined to agree and pick 2. So the changes thus far are:

- Use select to pick the right tool, using bazel-tools
- Mangle the path for Windows

Right?

For the next error, can you get it to print the full command with --sandbox_debug or one of its friends?

**Comment by amit-traiana** [comment]
> Use select to pick the right tool, using bazel-tools
> Mangle the path for Windows

Exactly. I'll create a Fork and link it so you can review those. 

> For the next error, can you get it to print the full command with --sandbox_debug or one of its friends?
Yes, I am using sandbox_debug/failure_verbose/subcommands. But it's a bit tricky on Windows because it runs MSYS to be able to run it as bash script. So it's not really CMD nor a real Bash. I'll try to debug it later on today :-)
Some machines around here are Windows, so we have no options but to support those.

**Comment by amit-traiana** [comment]
This is quite painful to debug. Let's start with the error message, which is quite obscure: 

` --grpc_scala_plugin_out: protoc-gen-grpc_scala_plugin: Plugin failed with status code 1.`

`protoc` outputs this error when the error code isn't `0`, but don't explain what exactly wen't wrong. So it's not very clear why it fails. I tried to run the command manually without Bazel, but this is also a problem. The full command is truncated:

`ERROR: C:/users/test/_bazel_automation/iesztftu/external/com_google_protobuf/BUILD:309:2: error executing shell command: 'C:/Users/Test/AppData/Local/Programs/Git/bin/bash.exe -c mkdir -p bazel-out/x64_windows-fastbuild/bin/external/com_google_protobuf/descriptor_proto/scala_grpc_compile_aspect_verb0 && bazel-ou...' failed (Exit 1)`

For some reason `verbose_failures` won't display the full message. It should. Looks like a Bazel bug in Windows. So I'm forced to use --subcommands, which outputs the following:

```
cd C:/users/automation/_bazel_automation/iesztftu/execroot/depo
C:/Users/Test/AppData/Local/Programs/Git/bin/bash.exe -c mkdir -p bazel-out/x64_windows-fastbuild/bin/external/com_github_scalapb_scalapb_repo/scalapb_proto/scala_grpc_compile_aspect_verb0 && bazel-out/host/bin/external/com_google_protobuf/protoc.exe $@  --descriptor_set_in=bazel-out/x64_windows-fastbuild/bin/external/com_google_protobuf/descriptor_proto-descriptor-set.proto.bin;bazel-out/x64_windows-fastbuild/bin/external/com_github_scalapb_scalapb_repo/scalapb_proto-descriptor-set.proto.bin --plugin=protoc-gen-grpc_scala_plugin=external\com_github_scalapb_scalapb\bin\protoc-gen-scala.bat --grpc_scala_plugin_out=grpc:bazel-out/x64_windows-fastbuild/bin/external/com_github_scalapb_scalapb_repo/scalapb_proto/scala_grpc_compile_aspect_verb0/scalapb_proto.jar protobuf/scalapb/scalapb.proto
```

That's a you mentioned, is not the 'real' command, as it includes the `$@` in it. So I tried to tweak the command to make it run, and after many attempts playing with quotes, I managed to run it successfully:

```
C:/Users/Test/AppData/Local/Programs/Git/bin/bash.exe -c "mkdir -p 'bazel-out/x64_windows-fastbuild/bin/cme/orgmodel/v2/organization_proto/scala_grpc_compile_aspect_verb0' && bazel-out/host/bin/external/com_google_protobuf/protoc.exe --descriptor_set_in='bazel-out/x64_windows-fastbuild/bin/external/com_google_protobuf/descriptor_proto-descriptor-set.proto.bin;bazel-out/x64_windows-fastbuild/bin/external/com_github_scalapb_scalapb_repo/scalapb_proto-descriptor-set.proto.bin' --plugin=protoc-gen-grpc_scala_plugin='external\com_github_scalapb_scalapb\bin\protoc-gen-scala.bat' --grpc_scala_plugin_out=grpc:'bazel-out/x64_windows-fastbuild/bin/external/com_github_scalapb_scalapb_repo/scalapb_proto/scala_grpc_compile_aspect_verb0/scalapb_proto.jar' protobuf/scalapb/scalapb.proto"
```

But I can't seems to make it work with Bazel, even when trying to reproduce the above in it. I can't really control the quotes and the start/end of the command, I'm not sure if Bazel add quotes or not, and also - adding the single quote around the parameters like I did before makes no chance. Still getting the same error.

Did you had a change to battle the Windows command line already and maybe have an idea? I found [this](https://github.com/bazelbuild/bazel/issues/3889) which mentions in Windows it's better to use `run` rather than `run_shell` like it currently use.

**Comment by aaliddell** [comment]
The reason `run_shell` is currently used is so that we can print debug info if you set `verbose = <1-4>` on your `<lang>_<proto|grpc>_<library|compile>` target. Using `verbose = 4` should print the state of the sandbox and the commands it's going to run. I guess on windows we could disable that ability and use `run` instead.

Have you pushed the code you've got up to now somewhere? I don't immediately have access to a machine with Windows and Bazel, but can probably setup a VM and test things out.

**Comment by amit-traiana** [comment]
I'm using `verbose = 4`, but sadly that doesn't indicated how `run_shell` decides to run the command. The command (after inspecting the `command` and `args` variables) runs perfectly from `CMD`. The problems starts when trying to run it inside a `bash` process due to `run_shell`. There's where it breaks down. Changing the slash type? that's only happening because `bash` runs the `protoc` command. I'm assuming that if `run` will be used, that hack will be irrelevant. 

I opened a [fork](https://github.com/amit-traiana/rules_proto_grpc/commits/master) for you to try it. It includes all the changes we talked about. I'm interesting to know if you get the full Bazel command using `verbose_failure` on Windows, because I can't (and If you suffer from the same problem, I will open a bug on the official Repo).

**Comment by amit-traiana** [comment]
All right. I figured out the last command-line quirks and it's now working. Will push changes into the Fork. However, the plug-in fails because it expects to have `$JAVA_HOME`. I have `$JAVA_HOME` defined on the Windows machine, and when I try to `bash -c echo $JAVA_HOME` I do get a current folder. But I'm assuming that sandbox works a bit different because when I set `verbose=4` and you print the `env` - I don't see `$JAVA_HOME` defined. 

To make sure it's the only problem, I modified `run_shell` with:

`env = { "JAVA_HOME": "/bla/folder" }`

And I can confirm I was finally able to build Scala gRPC code in Windows :-) Obviously, that's not a solution. Does `rules_proto_grpc` has some kind of way to handle language specific env variables?

**Comment by aaliddell** [comment]
Wow this is getting deep down the rabbit hole.

> Does rules_proto_grpc has some kind of way to handle language specific env variables? 

Not presently, although I guess it could go on the `ProtoPluginInfo` provider. Given that the env var is not going to be stable across users, it'd have to be the `use_default_shell_env` option instead. Does running with that option set to true work?

Digging into the `scalapbc.bat` file (I can't link to it as it doesn't appear to be in their repo...but it's in the release zip), it looks like we are triggering the `if "%JAVAOK%"=="false"` condition. This block ends with `exit /B 1`, which I guess is where you're getting your exit code 1 from earlier in this issue. This branch is triggered when it can't find a good java executable, suggesting perhaps Java is not being found in your standard path within the sandbox, hence the requirement for setting `JAVA_HOME`.

If you just open a terminal and run `java`, does it work as expected? If it does, does adding a random `java && ` to the beginning of the generated `run_shell` command in `aspect.bzl` work? Somewhere along the line something is failing to find Java and debugging this is diabolical. In theory, we would consume `JavaRuntimeInfo`, which [has the home available](https://docs.bazel.build/versions/master/skylark/lib/JavaRuntimeInfo.html#java_home), but I can't think of a simple way to do that only for the scala plugin.

**Comment by amit-traiana** [comment]
> Digging into the scalapbc.bat file (I can't link to it as it doesn't appear to be in their repo...but it's in the release zip), it looks like we are triggering the if "%JAVAOK%"=="false" condition. This block ends with exit /B 1, which I guess is where you're getting your exit code 1 from earlier in this issue.

Exactly. That's how I finally figured the issue. It was pain to catch it because while the batch file use `echo` for documentation, it also state `@echo off` at the very start - so there was no information why the plug-in fails. Diabolical debugging it was indeed :-)

> it'd have to be the use_default_shell_env option instead. Does running with that option set to true work?
> Edit: Ah I see in the fork that use_default_shell_env = True works
>  In theory, we would consume JavaRuntimeInfo, which has the home available, but I can't think of a simple way to do that only for the scala plugin.

I can confirm that the two Windows machines I tried, worked with my fork. 
Also, I can confirm `java` is in my machine path. If I'll just open `CMD` and run `java` - everything fine. It even works if I'll `bash -c java`. But when running from within the Sandbox, `java` can't be found (I double checked with `java &&` as you suggested).
I also think `use_default_shell_env` is the easiest way to go, but kinda scary because it's at the centre of things - and you don't know how it's going to effect people. For example - how the Java plug-in works without `$JAVA_HOME` or `java`? and what would happen if now is the user has `$JAVA_HOME` defined, and it's going to sip into the sandbox when building Java code? 

For the record, I tested it on my machine. But I don't own Mac for example, nor I tested all `grpc_x` rules. I'm assuming that if you fine with the current fix as stated in the fork, the next step is to run your tests on top of that. I'll open a pull request.

Thanks!
